onnx/defs/tensor/defs.cc

```diff
@@ -403,46 +403,59 @@ ONNX_OPERATOR_SET_SCHEMA(
           if (!hasNInputShapes(ctx, 1)) {
             return;
           }
+
+          const auto& shape = ctx.getInputType(0)->tensor_type().shape();
+          int rank = shape.dim_size();
+          int axis = static_cast<int>(getAttribute(ctx, "axis", 0));
+          if (axis < -rank || axis >= rank) {
+            fail_type_inference(
+                "Invalid value of attribute 'axis'. Rank=",
+                rank,
+                " Value=",
+                axis);
+          }
+          if (axis < 0) {
+            axis += rank;
+          }
+          const auto& splitDim = shape.dim(axis);
+          if (!splitDim.has_dim_value()) {
+            return;
+          }
+          int splitDimValue = static_cast<int>(splitDim.dim_value());
+
           std::vector<int64_t> split;
-          if (!getRepeatedAttribute(ctx, "split", split)) {
-            if (!ctx.getInputType(0)->tensor_type().has_shape()) {
-              return;
-            }
-            const auto& shape = ctx.getInputType(0)->tensor_type().shape();
-            int rank = shape.dim_size();
-            int axis = static_cast<int>(getAttribute(ctx, "axis", 0));
-            if (axis < -rank || axis >= rank) {
-              fail_type_inference(
-                  "Invalid value of attribute 'axis'. Rank=",
-                  rank,
-                  " Value=",
-                  axis);
+          if (getRepeatedAttribute(ctx, "split", split)) {
+            if (split.size() != ctx.getNumOutputs()) {
+              fail_shape_inference(
+                  "Mismatch between number of splits (",
+                  split.size(),
+                  ") and outputs (",
+                  ctx.getNumOutputs());
             }
-            if (axis < 0) {
-              axis += rank;
+            int64_t totalDim = 0;
+            for (int64_t d : split) {
+              totalDim += d;
             }
-            const auto& splitDim = shape.dim(axis);
-            if (!splitDim.has_dim_value()) {
+            if (totalDim != splitDimValue) {
               return;
             }
-            int splitDimValue = static_cast<int>(splitDim.dim_value());
+          } else {
             int chunkSize =
```

Comment:
Per spec, it says if 'split' is not provided, split it into equal sized parts, so should the case where splitDimValue % ctx.getNumOutputs() != 0 be an error condition ?

Reply:
Otherwise, we might have to arbitrarily deal with it like the logic in 444-448. The logic is not prescribed per spec. So I think this should be an error condition as well. To me there are 3 possible error conditions in this op-

1) When 'split' is present and its count does not match number of outputs
2) When 'split' is present and its sum does not match input shape's corresponding value in the 'axis' dimension
3) When 'split' is not present and  input shape's corresponding value in the 'axis' dimension cannot be split evenly to the number of outputs

Reply:
Done